Refactor get numerical jacobian to calculate wrt all outputs at once

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* #55656 Clean up raise exception logic
* #55237 Enable fast gradcheck for real inputs and outputs
* #54480 Implement faster gradcheck but not enabled for most things
* #54479 Factor out numerical logic
* **#54378 Refactor get numerical jacobian to calculate wrt all outputs at once**

### For release notes
`torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being deprecated. 

In the future, user code relying on this function will break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.

Differential Revision: [D27728720](https://our.internmc.facebook.com/intern/diff/D27728720)

soulitzer committed:
(WIP) Refactor numerical to calculate wrt all outputs at once

[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer commented on test/test_overrides.py:
```diff
@@ -802,32 +802,35 @@ def test_gradcheck(self):
             torch.autograd.grad,
             torch.add,
-        })
+        }, total_used_calls)
```
Flipped this around because it should be "expected" before "actual"

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor numerical to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

Other changes: 
- Refactors make_jacobian so that it can be used for fast gradcheck as well
- `iter_tensor` is simply the outer loop factored out into a generator, no logic should be changed tehre.

For next PR:
- Adds `fast_mode` arguments to both gradcheck and gradgradcheck (default False)
- Sets default for internal gradcheck wrapper to be True


[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -94,77 +113,100 @@ def get_stride(size):
-            return fn([x.to_mkldnn()])
+            inp = [x.to_mkldnn()]
+        elif layout == torch.sparse_coo:
+            inp = [a.clone() for a in _as_tuple(inputs)]
```
Temporary fix to address the issue detailed here: https://github.com/pytorch/pytorch/pull/52874. How it affects us here is that if we do `out = f(input)` and input is uncoalesced, and then we call `.coalesce()` on that input, the input will be moved off the out's backward graph, so that when we  do `.grad(out, input)` later we would always get zero (when allow_unused=True). Because we call `f` again, during get_numerical_jacobian, we may run into this issue if `f` calls `coalesce`. Note that this wasn't an issue previously because we looped through the inputs one by one and always compute numerical after we compute the analytical.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

zou3519 commented:
> This PR addresses the comment here #53857 (comment), to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here.

What exactly changed? Does this PR make it so that `func` only gets called 2 * num_inputs times?

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -19,21 +19,39 @@ def zero_gradients(x):
+    for t in output_tensors:
+        if is_float_or_complex_tensor(t):
+            if dim is None:
+                out.append(t.new_zeros((t.nelement(),), **options))
+            else:
+                out.append(t.new_zeros((dim, t.nelement()), **options))
```
nit: prefer numel() to nelement(), nelement() is one of those legacy APIs that we have

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -19,21 +19,39 @@ def zero_gradients(x):
+    return tuple(out)
+
+
+def make_jacobians_with_outputs(output_tensors: Tuple, dtype=None, device=None, dim=None):
```
Why have a `dim=None`? It looks like the only callsite uses `dim=input.numel()`. Ditto for `make_jacobians_with_inputs`.

Also nit: I always thought "make_jacobian" was a nondescriptive name -- we're not making the jacobian, we're just allocating memory for it.

soulitzer replied:
It had dim=None because I wanted to prepare for the fast gradcheck version, but I guess I should just introduce it there instead. make_jacobians_with_inputs turned out to not need the dim=None case at all though even in the fast case, so removing that entirely.

I'm updating it to be `allocate_jacobians_with_x`

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -493,6 +547,14 @@ def get_notallclose_msg(analytical, numerical, output_idx, input_idx, error_str=
     return error_str + 'Jacobian mismatch for output %d with respect to input %d,\n' \
         'numerical:%s\nanalytical:%s\n' % (output_idx, input_idx, numerical, analytical)
 
+def transpose(m):
```
nit: m -> matrix_of_tensors. I needed to think for a bit about what was actually getting transposed

Also, there's a one-liner for this (that is slightly different, returns a List[Tuple[Tensor]] if that matters)
```
def transpose(m):
    return zip(*m)
```

soulitzer replied:
So clean!

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -93,76 +111,114 @@ def get_stride(size):
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3, grad_out=1.0):
+    """Computes the numerical jacobian for a given fn and inputs. Returns M * N jacobians
+    where M is the number of input tensors that require grad, and N is the number of output
+    float/complex tensors.
```
Should add the type hint for `List[Tuple[torch.Tensor]]` and make it clear in the docstring that we're returning M lists of N tuples of jacobians

zou3519 commented on test/test_autograd.py:
```diff
@@ -4195,6 +4198,26 @@ def fn3(x):  # C -> R
+        def fn(x, y):
+            return x + y.to_dense()
+        a = torch.rand(10, requires_grad=True)
+        b = torch.rand(10, dtype=torch.float32).to_mkldnn().requires_grad_(True)
+        self.assertTrue(gradcheck(fn, (a, b), atol=1e-1, check_batched_grad=False))
+
+        def fn2(x, y):
+            return x.to_dense() + y.to_dense()
+        c = torch.rand(10, dtype=torch.float32).to_mkldnn().requires_grad_(True)
+        self.assertTrue(gradcheck(fn, (a, c), atol=1e-1, check_batched_grad=False))
```
These atols are very lenient, is that expected?

soulitzer replied:
> > This PR addresses the comment here #53857 (comment), to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here.
> 
> What exactly changed? Does this PR make it so that `func` only gets called 2 * num_inputs times?
> 
> > However, because very few ops actually return multiple outputs
> 
> Are we talking about number of output tensors, or the total numel of all output tensors?

Oh I think there was a newline in the original comment that cut the last part off. A more precise way of saying it should be: New: `2 * total_number_of_elements_in_all_the_inputs * num_outputs`
Old: `2 * total_number_of_elements_in_all_the_inputs`
Fast gradcheck: `2`

This is because before we had this line here https://github.com/pytorch/pytorch/pull/54378/files#diff-440fd3cf8800df054b6c18370faac5566e0ee1f14cc6c8d93d6f436b37dfabc3L302
where we restrict ourselves to utilize only a single output at a time. The way it was done previously is that we looped through both outputs and inputs in a doubly nested loop, and inside this we compute the Jacobian corresponding to input i and output j, which required calling f `2 * `input_i.numel` times. 

After this PR, we utilize the entire potential of calling `f`, so we only loop through the inputs. And since calling f returns all the outputs, we can compute all Jacobians corresponding to input i with the same 2 * `input_i.numel` calls to f.

soulitzer replied:
Oh wait I just checked and the actual and expected actually differ by 1.33e-2, so we'd indeed need an atol of 1e-1 (or at least 2e-2).

zou3519 reviewed (commented):
still reading through it, but here are some initial comments

soulitzer commented:
> > This PR addresses the comment here #53857 (comment), to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here.
> 
> What exactly changed? Does this PR make it so that `func` only gets called 2 * num_inputs times?
> 
> > However, because very few ops actually return multiple outputs
> 
> Are we talking about number of output tensors, or the total numel of all output tensors?

Oh I think there was a newline in the original comment that cut the last part off. A more precise way of saying it should be: New: `2 * total_number_of_elements_in_all_the_inputs * num_outputs`
Old: `2 * total_number_of_elements_in_all_the_inputs`
Fast gradcheck: `2`

This is because before we had this line here https://github.com/pytorch/pytorch/pull/54378/files#diff-440fd3cf8800df054b6c18370faac5566e0ee1f14cc6c8d93d6f436b37dfabc3L302
where we restrict ourselves to utilize only a single output at a time. The way it was done previously is that we looped through both outputs and inputs in a doubly nested loop, and inside this we compute the Jacobian corresponding to input i and output j, which required calling f `2 * `input_i.numel` times. 

After this PR, we utilize the entire potential of calling `f`, so we only loop through the inputs. And since calling f returns all the outputs, we can compute all Jacobians corresponding to input i with the same 2 * `input_i.numel` calls to f.

soulitzer commented:
TODO: Add checks similar to what we have for analytical jacobian to verify shape/type is correct for the forward pass as discussed with @albanD offline.

This should be the correct PR in the stack to introduce this since the previous and next ones shouldn't modify behavior.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+        return input
+
+
+def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, delta):
```
@nikitaved Hey! Heard from @albanD that adding such a check in gradcheck might be useful (based on an issue you guys encountered recently). I was wondering if you had any more details of such a usecase, and whether what I'm adding covers it.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

zou3519 reviewed (commented):
Some more questions and comments about what the code is doing after the refactor. A lot of the readability problems seems pre-existing since the old code is difficult to read as well (there are functions with a lot of inputs and the names of the functions don't quite say what they do).

Hopefully after some Q&A I'll understand the implementation and can finally give a full review

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+        else:
+            return input
+    elif input.layout == torch.sparse_coo:
+        # modifications to entry are reflected in input so we could've just returned `input` here
+        # but due to an issue with coalesce, we need to do an extra clone here.
+        # TODO: get rid of this extra clone once https://github.com/pytorch/pytorch/pull/52874 is landed
+        # Make this new tensor require again in case the function has hooks
+        return torch.sparse_coo_tensor(input._indices(), input._values(), input.size()).requires_grad_(True)
```
There is an issue where calling .coalesce on a tensor moves it off the graph when the tensor is already coalesced, so analytical would always return 0 wrt to that input if it is previously used to compute forward pass. Previously this wasn't an issue because we always computed the analytical jacobian before numerical jacobian for any particular output (and we probably didn't have any tests where inputs and sparse and where we return multiple outputs which would've also surfaced the issue). But as part of the refector we now compute numerical jacobians before we compute analytical jacobians. 
Also, I think @peterbell10 was working on a PR fixing this though, so once that lands a lot of our logic concerning sparse would be no longer needed.

peterbell10 replied:
@soulitzer the coalesce fix (#52874) was just merged BTW.

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
-        else:
-            # convert the dense tensor back to have mkldnn layout
-            return fn([x.to_mkldnn()])
+        inp = tuple(prepped_input(a, i, x, input_idx) if is_tensor_like(a) else a
+                    for i, a in enumerate(_as_tuple(inputs)))
+        return tuple(a.clone() for a in _as_tuple(fn(*inp)))
```
Whoops I should've mentioned this in the PR description. This PR fixes some issues in the original implementation. For example: the original implementation did not allow for multiple inputs when any of the inputs are mkldnn, and also theres that same issue regarding sparse here https://github.com/pytorch/pytorch/pull/54378/files#r604493746.

zou3519 replied:
Aha, that makes more sense now!

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+                    #            = real(grad_out.conj() * ds_dx)
+                    d[d_idx] = torch.real(grad_out.conjugate() * ds_dx)
+                else:   # R -> R
+                    # skip if grad_out is complex but output is real
+                    if not isinstance(grad_out, complex):
+                        d[d_idx] = ds_dx * grad_out
```
Previously, we looped through outputs, then input. In the refactor we flip this around so we loop through inputs first, then outputs. The issue is that we should apply a grad_out of 1j if the outputs is complex, so the logic that was previously here: https://github.com/pytorch/pytorch/blob/887c9c768f0f2df61b4a0b43931018d4aec07487/torch/autograd/gradcheck.py#L603

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+    return jacobians
+
+
+def prepped_input(input, input_idx, entry, entry_idx):
+    # Prepares the inputs to be passed into the function while including the new modified input.
```
The way its called is that instead of doing something like `f(tupled_inputs)`, we first wrap it in `f([prepped_input(a, i, entry, entry_idx) for i, a in enumerate(tupled_inputs))]`. `entry` is the particular input was is perturbed for the calculation (since we want to compute the numerical jacobian wrt to that particular input). `input` is the `a` that we happen to loop to. Depending on the layout of input, and whether `input_idx` is equal to `entry_idx` we may want to "prepare" it first before passing it back to `f`.

For example if we have some f taking a, b, where a is strided and b is mkldnn. We we need to compute derivatives wrt b, we need to 1) use the entry we modified instead of input because entry is a clone so changes in entry are not reflected in input 2) we should convert it back to mkldnn with .to_mkldnn because f expects b to be mkldnn. 

Conversely, we cannot always simply return entry, because in the strided case for example, entry is input.data, which is not part of the original graph, and this would result in a failure for the gradgrad case.

zou3519 replied:
I think what really threw me off here was looking at `input_idx` and `entry_idx` and wondering when they're different or the same. Reading the callsite of prepped_input helped that understanding.

What if we changed this function to:
```
prepped_input(input, maybe_perturbed_input)
```
?
`maybe_perturbed_input` is the perturbed input if it corresponds to this input, otherwise it is None. So we'd change `fn_out` to be:
```
    def fn_out():
        inp = tuple(prepped_input(input, x if i == input_idx else None) if is_tensor_like(input) else input
                    for i, input in enumerate(_as_tuple(inputs)))
        return tuple(a.clone() for a in _as_tuple(fn(*inp)))
```

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
 
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
```
Hmm interesting. The one big repo that I see using this is  https://github.com/NVIDIA/MinkowskiEngine/blob/b40b98d2ee2f3f18c51ac609c0179482ffac071f/MinkowskiEngine/utils/gradcheck.py. 
Maybe we can just file an issue there.

I also see people using `get_analytical_jacobian` which we've already removed. So not sure if its worth it to maintain some kind of unofficial stable api for this type of thing since we'd have to also recreate how the old function returned `reentrant`, `grad_correct_dtype` etc.

soulitzer replied:
Updated the PR description to have a bc-breaking description. Also updated here as well https://github.com/pytorch/pytorch/pull/54049.

zou3519 replied:
Offline Alban expressed that this is reasonable but also asked if there's any way we can keep the old API lying around. We could potentially copy & paste the old API (since it's all python files) and have it go through a deprecation cycle?

soulitzer replied:
@zou3519 I added some wrapper functions that replicated the old API's behavior. The `get_numerical_jacobian` we use internally is now `_get_numerical_jacobian`. How does the deprecation message look?

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+         f" dtypes {output1.dtype} and {output2.dtype}.")
+
+
+def compute_gradient(fn, inputs, input_idx, x, idx, delta, eps, layout):
```
Yeah its a bit confusing for sure. `idx` here indexes into `x`, which isn't exactly the same as inputs[input_idx]. You can check the iter_tensor code which handles the input -> x logic. 

FYI though, this gets immediately refactored in the next PR which I think makes this function a lot cleaner.

zou3519 reviewed (commented):
Code LGTM to me. Added some suggestions to help the code read better (but those aren't blocking).

We should figure out what to do with all the users who are calling get_numeric_jacobian and get_analytic_jacobian. Filing issues to their repos as you suggested sounds reasonable; I think we should tag this PR as BC-breaking and add a description to be included into the release notes

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,132 @@ def get_stride(size):
+        target = inputs
+    inp_indices = [i for i, a in enumerate(target) if is_tensor_like(a) and a.requires_grad]
+    for i, (inp, inp_idx) in enumerate(zip(iter_tensors(target, True), inp_indices)):
+        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
```
Oh I just realized that this also goes away in the next PR, so I'm just going to make the update there.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer commented:
Hey @zou3519, I've added a small note on the bc-breakingness of this PR. I didn't include an example workaround here as the guidelines suggest because the I think the proper workaround here should really just be to not rely on the non-public api. 

I also created an issue https://github.com/NVIDIA/MinkowskiEngine/issues/341 warning the upcoming breakage.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -4196,6 +4199,57 @@ def fn3(x):  # C -> R
+        with self.assertRaisesRegex(AssertionError, 'return outputs with the same dtype when inputs are perturbed'):
+            self.assertTrue(gradcheck(fn2, (a,)))
+
+    def test_gradcheck_complex_non_complex_outputs(self):
```
This was the test that was added.

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,148 @@ def get_stride(size):
 
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def _get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
```
We should probably prefix all other helper functions in this file with `_` so that no one tries to rely on them. An alternative could be to add them to some `torch.autograd._gradcheck` file that has an underscore prefix.

soulitzer replied:
Created an issue for this here https://github.com/pytorch/pytorch/issues/55714. It would take many times the effort to have to modify them names here due to all the rebase conflicts.

zou3519 replied:
Yeah, as long as we can do this before 1.9 it doesn't matter when it happens

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,148 @@ def get_stride(size):
+
+def get_numerical_jacobian(fn, inputs, target=None, eps=1e-3, grad_out=1.0):
+    # Simple wrapper around _get_numerical_jacobian
+    warnings.warn("get_analytical_jacobian is deprecated!")
```
Since we're doing a deprecation, let's remove the BC-breaking label and add a "deprecation" label (sorry for the thrashing!)

soulitzer replied:
No worries!

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,148 @@ def get_stride(size):
+
+def get_numerical_jacobian(fn, inputs, target=None, eps=1e-3, grad_out=1.0):
+    # Simple wrapper around _get_numerical_jacobian
+    warnings.warn("get_analytical_jacobian is deprecated!")
```
Let's make this deprecation message more descriptive. It's good for deprecation messages to include a reason and a timeline.

So, something like:
```
get_numerical_jacobian was part of PyTorch's private API and not meant to be exposed.
We are deprecating it and it will be removed in a future version of PyTorch. If you have a
specific use for this or feature request for this to be a stable API, please file us an issue
at <link_to_issue_tracker>
```
Ditto for `get_analytical_jacobian`

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -240,6 +324,25 @@ def fail_test(msg):
     return jacobians, failed
 
 
+def get_analytical_jacobian(inputs, output, nondet_tol=0.0, grad_out=1.0):
+    # Replicates the behavior of the old get_analytical_jacobian before the refactor
+    warnings.warn("get_analytical_jacobian is deprecated!")
+    diff_input_list = list(iter_tensors(inputs, True))
+
+    def backward_fn(grad_output):
+        return torch.autograd.grad(output, diff_input_list, grad_output,
+                                   retain_graph=True, allow_unused=True)
+
+    jacobians_rows = compute_analytical_jacobian_rows(backward_fn, output.clone(), grad_out)
+    jacobians_rows_reentrant = compute_analytical_jacobian_rows(backward_fn, output.clone(), grad_out)
+
+    jacobians, correct_grad_types, correct_grad_sizes = combine_jacobian_rows(jacobians_rows, inputs, output)
+    jacobians_reentrant, _, _ = combine_jacobian_rows(jacobians_rows_reentrant, inputs, output)
+    reentrant = check_jacobians_equal(jacobians, jacobians_reentrant, nondet_tol)
+
+    return jacobians, reentrant, correct_grad_sizes, correct_grad_types
```
Since we're not using `get_analytical_jacobian` anymore in our testing code, could we add a correctness test for it to make sure that it still works? Ditto for `get_numerical_jacobian`.

It would also be nice to test that the deprecation message shows up properly.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,148 @@ def get_stride(size):
+
+
+def get_numerical_jacobian(fn, inputs, target=None, eps=1e-3, grad_out=1.0):
+    # Simple wrapper around _get_numerical_jacobian
```
We should probably keep around the old docstring: https://github.com/pytorch/pytorch/blob/55d45458bdb0b404bcaba540978123cbfccfbefa/torch/autograd/gradcheck.py#L130-L137

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,148 @@ def get_stride(size):
+    def fn_pack_inps(*inps):
+        return fn(inps)
+    jacobians = _get_numerical_jacobian(fn_pack_inps, inputs, None, target, eps, grad_out)
+    return jacobians[0][0]
```
`jacobians[0][0]` is a single Tensor, right? The old code (https://github.com/pytorch/pytorch/blob/58eb23378f2a376565a66ac32c93a316c45b6131/torch/autograd/gradcheck.py#L49) looks like it can return a list of tensors if `input` is a list of tensors. Is that correct?

soulitzer replied:
Oh right, good catch. Looks like it always returned a tuple of Tensors before. I've updated it to be `tuple(jacobian_for_each_output[0] for jacobian_for_each_output in jacobians)`

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### BC Breaking Notes (not part of public api):
The api for `torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being updated. User code relying on this function will now break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end bc-breaking notes --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor get numerical jacobian to calculate wrt all outputs at once"


### For release notes
`torch.autograd.gradcheck.get_numerical_jacobian` (not part of the public api) is being deprecated. 

In the future, user code relying on this function will break because, among other changes, `get_numerical_jacobian` now returns `List[Tuple[torch.Tensor]]` instead of `List[torch.Tensor]`.

(more details if necessary)
For a `fn` that takes in M inputs and N outputs we now return a list of M N-tuples of jacobians where `output[i][j]` would represent the numerical jacobian w.r.t. to the ith input and the jth output. Previously `get_numerical_jacobian` returned a list of tensors where each tensor represents the jacobian w.r.t. to each of the M inputs and a specific output. Finally, the function passed in as the parameter `fn` should expect to handle individual parameters, where previously `fn` is required to expect its parameters wrapped in a tuple.

--- end --

This PR addresses the comment here https://github.com/pytorch/pytorch/pull/53857#discussion_r595429639, to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here. 

The main benefit of doing this change as part of the refactor is that it helps us isolate the possible bugs that are specific to switching `get numerical jacobian` to run in a per output way vs all outputs at once. Much of the logic implemented here will be the same for the fast gradcheck case, so knowing for certain that everything should pass after this stage will make the next step much simpler.

The get_numerical_jacobian api is also being used in common_nn. So we update the callsite there as well.


[ghstack-poisoned]

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -102,76 +114,166 @@ def get_stride(size):
+        return fn(inps)
+    jacobians = _get_numerical_jacobian(fn_pack_inps, inputs, None, target, eps, grad_out)
+
+    return tuple(jacobian_for_each_output[0] for jacobian_for_each_output in jacobians)
```
Updated this here: since old API always returned a tuple of tensors. Tensors represent jacobians wrt to each input and to a single output.

soulitzer commented on test/test_autograd.py:
```diff
@@ -4157,6 +4160,110 @@ def fn3(x):  # C -> R
+        c = torch.ones(2, 2, requires_grad=True, dtype=torch.complex128)
+        self.assertTrue(gradcheck(fn2, (c)))
+
+    def test_gradcheck_get_numerical_jacobian(self):
```
Added tests for get_numerical_jacobian and get_analytical_jacobian

zou3519 reviewed (approved):
lgtm!
